### ISSUE TITLE
ERNIE 论文中的完形填空测试应该怎么做？有人做过吗？

### ISSUE THREAD

**Issue by Finley1991** [PaddlePaddle/ERNIE#173]
最近看到百度的大神们和各位小伙伴都在测试百度ERNIE和BERT中文模型的效果，其中有一项完形填空测试，请问这个测试是怎么做到的？ERNIE模型中就有这个接口么？哪位大神能帮忙解释一下？（我已经了解到ERNIE的训练过程 mask LM任务应该是可以实现这个功能，可是问题是我找不到调用的借口和方法）

**Comment by nbcc** [comment]
我们通过完全填空类的任务，证明 ERNIE 语义建模方式较 BERT 的优势。具体的调用的接口，你可以依照开源 mask lm 的代码改造一下。

**Comment by ddonng** [comment]
可否给个链接……

**Comment by nbcc** [comment]
> 可否给个链接……

样本可以直接生成带 mask 的输入，关闭生成 mask 逻辑
https://github.com/PaddlePaddle/LARK/blob/f2683894a159ed5eae19113927f4a191e9f50f93/ERNIE/batching.py#L60

预测
https://github.com/PaddlePaddle/LARK/blob/f2683894a159ed5eae19113927f4a191e9f50f93/ERNIE/model/ernie.py#L206

**Comment by jfkey** [comment]
> 最近看到百度的大神们和各位小伙伴都在测试百度ERNIE和BERT中文模型的效果，其中有一项完形填空测试，请问这个测试是怎么做到的？ERNIE模型中就有这个接口么？哪位大神能帮忙解释一下？（我已经了解到ERNIE的训练过程 mask LM任务应该是可以实现这个功能，可是问题是我找不到调用的借口和方法）
请问你弄明白了完形填空是如何做的了么？代码在上述所给链接有解决问题么？您能简要的说一下么？

**Comment by jfkey** [comment]
> > 可否给个链接……
> 
> 样本可以直接生成带 mask 的输入，关闭生成 mask 逻辑
> https://github.com/PaddlePaddle/LARK/blob/f2683894a159ed5eae19113927f4a191e9f50f93/ERNIE/batching.py#L60
> 
> 预测
> https://github.com/PaddlePaddle/LARK/blob/f2683894a159ed5eae19113927f4a191e9f50f93/ERNIE/model/ernie.py#L206

请问Ernie中的完形填空的预测任务，是不是从vocab中选取应该填在空格中的词语？

**Comment by playedcreat** [comment]
有没有大佬实现成功的，指教一下这个应该怎么弄吗，我还是不太会。。

**Comment by Meiyim** [comment]
在最新开源的[ERNIE动态图模型](https://github.com/PaddlePaddle/ERNIE/tree/dygraph)当中，做完形填空将会更加简单方便.

继承`ErnieModelForPretraining`并改写`forward`方法，便可以实现简单的完形填空逻辑(仅限ernie1.0中文模型；enire2.0等英文模型因为没有发布mask_lm任务相关的weight文件所以还无法做到完形填空)；

请参考：
```python
import os
import numpy as np

import paddle.fluid as F
import paddle.fluid.layers as L
import paddle.fluid.dygraph as D
from ernie.modeling_ernie import ErnieModelForPretraining, ErnieModel
from ernie.tokenizing_ernie import ErnieTokenizer

class ErnieCloset(ErnieModelForPretraining):
    def __init__(self, *args, **kwargs):
        super(ErnieCloset, self).__init__(*args, **kwargs)
        del self.pooler_heads
    def forward(self, src_ids, *args, **kwargs):
        pooled, encoded = ErnieModel.forward(self, src_ids, *args, **kwargs)
        encoded_2d = L.gather_nd(encoded, L.where(src_ids == mask_id))
        encoded_2d = self.mlm(encoded_2d)
        encoded_2d = self.mlm_ln(encoded_2d)
        logits_2d = L.matmul(encoded_2d, self.word_emb.weight, transpose_y=True) + self.mlm_bias
        return logits_2d


if __name__ == "__main__":
    model_dir = 'ernie-1.0'

    D.guard().__enter__()
    tokenizer = ErnieTokenizer.from_pretrained(model_dir)
    rev_dict = {v: k for k, v in tokenizer.vocab.items()}
    rev_dict[tokenizer.pad_id] = '' # replace [PAD]
    rev_dict[tokenizer.sep_id] = '' # replace [PAD]
    rev_dict[tokenizer.unk_id] = '' # replace [PAD]

    @np.vectorize
    def rev_lookup(i):
        return rev_dict[i]

    ernie = ErnieCloset.from_pretrained(model_dir)
    ernie.eval()

    ids, _ = tokenizer.encode('戊戌变法，又称百日维新，是 [MASK] [MASK] [MASK] 、梁启超等维新派人士通过光绪帝进行 的一场资产阶级改良。')
    mask_id = tokenizer.mask_id
    print(ids)
    ids = np.expand_dims(ids, 0)
    ids = D.to_variable(ids)
    logits = ernie(ids).numpy()
    output_ids = np.argmax(logits, -1)
    seg_txt = rev_lookup(output_ids)
    print(seg_txt)
```

**Comment by ouhongxu** [comment]
> 在最新开源的[ERNIE动态图模型](https://github.com/PaddlePaddle/ERNIE/tree/dygraph)当中，做完形填空将会更加简单方便.
> 
> 继承`ErnieModelForPretraining`并改写`forward`方法，便可以实现简单的完形填空逻辑(仅限ernie1.0中文模型；enire2.0等英文模型因为没有发布mask_lm任务相关的weight文件所以还无法做到完形填空)；
> 
> 请参考：
> 
> ```python
> import os
> import numpy as np
> 
> import paddle.fluid as F
> import paddle.fluid.layers as L
> import paddle.fluid.dygraph as D
> from ernie.modeling_ernie import ErnieModelForPretraining, ErnieModel
> from ernie.tokenizing_ernie import ErnieTokenizer
> 
> class ErnieCloset(ErnieModelForPretraining):
>     def __init__(self, *args, **kwargs):
>         super(ErnieCloset, self).__init__(*args, **kwargs)
>         del self.pooler_heads
>     def forward(self, src_ids, *args, **kwargs):
>         pooled, encoded = ErnieModel.forward(self, src_ids, *args, **kwargs)
>         encoded_2d = L.gather_nd(encoded, L.where(src_ids == mask_id))
>         encoded_2d = self.mlm(encoded_2d)
>         encoded_2d = self.mlm_ln(encoded_2d)
>         logits_2d = L.matmul(encoded_2d, self.word_emb.weight, transpose_y=True) + self.mlm_bias
>         return logits_2d
> 
> 
> if __name__ == "__main__":
>     model_dir = 'ernie-1.0'
> 
>     D.guard().__enter__()
>     tokenizer = ErnieTokenizer.from_pretrained(model_dir)
>     rev_dict = {v: k for k, v in tokenizer.vocab.items()}
>     rev_dict[tokenizer.pad_id] = '' # replace [PAD]
>     rev_dict[tokenizer.sep_id] = '' # replace [PAD]
>     rev_dict[tokenizer.unk_id] = '' # replace [PAD]
> 
>     @np.vectorize
>     def rev_lookup(i):
>         return rev_dict[i]
> 
>     ernie = ErnieCloset.from_pretrained(model_dir)
>     ernie.eval()
> 
>     ids, _ = tokenizer.encode('戊戌变法，又称百日维新，是 [MASK] [MASK] [MASK] 、梁启超等维新派人士通过光绪帝进行 的一场资产阶级改良。')
>     mask_id = tokenizer.mask_id
>     print(ids)
>     ids = np.expand_dims(ids, 0)
>     ids = D.to_variable(ids)
>     logits = ernie(ids).numpy()
>     output_ids = np.argmax(logits, -1)
>     seg_txt = rev_lookup(output_ids)
>     print(seg_txt)
> ```

可以提供一个返回前n个适宜词的方法吗？万分感谢~！
上面方法仅返回一个答案，而且没有词检验（如果一个填空中包含两个[MASK]，第一个mask取字典中概率最高的那个字，第二个mask也取字典中概率最高的那个字，有时这两个mask拼接起来不是一个词）。
